releng/org.eclipse.n4js.releng.utils/src/org/eclipse/n4js/antlr/N4JSAntlrGeneratorFragment2.xtend

```diff
@@ -96,12 +120,127 @@ class N4JSAntlrGeneratorFragment2 extends N4AntlrGeneratorFragment2 {
+		file.content = '''
+			public class «grammar.parserClass.simpleName» extends «grammar.getParserSuperClass(partialParsing)» {
+			
+				@«Inject»
+				private «grammar.grammarAccess» grammarAccess;
+			
+				private «Map»<«AbstractElement», String> nameMappings;
+			
+				@Override
+				protected «grammar.internalParserClass» createParser() {
+					«grammar.internalParserClass» result = new «grammar.internalParserClass»(null);
+					result.setGrammarAccess(grammarAccess);
+					return result;
+				}
+			
+				«IF hasSyntheticTerminalRule»
+					@Override
+					protected «TokenSource» createLexer(«CharStream» stream) {
+						return new «grammar.tokenSourceClass»(super.createLexer(stream));
+					}
+					
+				«ENDIF»
+				@Override
+				protected String getRuleName(«AbstractElement» element) {
+					if (nameMappings == null) {
+						«grammar.initNameMappings()»
+					}
+					return nameMappings.get(element);
+				}
+			
+				@Override
+				protected «Collection»<«"org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElement".typeRef»> getFollowElements(«grammar.internalParserSuperClass» parser) {
+					try {
+						«grammar.internalParserClass» typedParser = («grammar.internalParserClass») parser;
+						typedParser.«grammar.allParserRules.head.originalElement.entryRuleName»();
+						return typedParser.getFollowElements();
+					} catch(«RecognitionException» ex) {
+						throw new «RuntimeException»(ex);
+					}
+				}
+			
+				@Override
+				protected String[] getInitialHiddenTokens() {
+					return new String[] { «FOR hidden : grammar.initialHiddenTokens SEPARATOR ", "»"«hidden»"«ENDFOR» };
+				}
+			
+				public «grammar.grammarAccess» getGrammarAccess() {
+					return this.grammarAccess;
+				}
+			
+				public void setGrammarAccess(«grammar.grammarAccess» grammarAccess) {
+					this.grammarAccess = grammarAccess;
+				}
+			}
+		'''
+		file
+	}
+
+	/**
+	 * Produce the initial name mappings for the grammar. Handles parameterized rule calls.
+	 */
+	private def StringConcatenationClient initNameMappings(Grammar it) {
+		val RuleFilter filter = new RuleFilter();
+		filter.discardUnreachableRules = true
+		val RuleNames ruleNames = RuleNames.getRuleNames(it, true);
+		val Grammar flattened = new FlattenedGrammarAccess(ruleNames, filter).getFlattenedGrammar();
+		val Set<AbstractElement> seenElements = newHashSet
+		return '''
+			«ImmutableMap».Builder<«AbstractElement», String> nameMappingsBuilder = «ImmutableMap».builder();
+			«FOR element : (flattened.allAlternatives + flattened.allGroups + flattened.allAssignments + flattened.allUnorderedGroups).filter(AbstractElement).filter[seenElements.add(originalElement)]»
+				nameMappingsBuilder.put(grammarAccess.«element.originalElement.grammarElementAccess», "«element.originalElement.containingRule.contentAssistRuleName»__«element.originalElement.gaElementIdentifier»«IF element instanceof Group»__0«ENDIF»");
+			«ENDFOR»
+			nameMappings = nameMappingsBuilder.build();
+		'''
+	}
+
```

Comment:
The above (beginning at your comment) will be removed with 2.13. Please a short marker here to mark the end of the code to be deleted.